blmath/geometry/primitives/plane.py

```diff
@@ -262,103 +264,131 @@ def mesh_xsections(self, m, neighborhood=None):
 
         Returns a list of Polylines.
         '''
-        import operator
         from blmath.geometry import Polyline
 
         # 1: Select those faces that intersect the plane, fs
-        sgn_dists = self.signed_distance(m.v)
-        which_fs = np.abs(np.sign(sgn_dists)[m.f].sum(axis=1)) != 3
-        fs = m.f[which_fs]
-
+        fs = self.mesh_intersecting_faces(m)
         if len(fs) == 0:
             return [] # Nothing intersects
+        # and edges of those faces
+        es = np.vstack((fs[:, (0, 1)], fs[:, (1, 2)], fs[:, (2, 0)]))
 
         # 2: Find the edges where each of those faces actually cross the plane
-        def edge_from_face(f):
-            face_verts = [
-                [m.v[f[0]], m.v[f[1]]],
-                [m.v[f[1]], m.v[f[2]]],
-                [m.v[f[2]], m.v[f[0]]],
-            ]
-            e = [self.line_segment_xsection(a, b) for a, b in face_verts]
-            e = [x for x in e if x is not None]
-            return e
-        edges = np.vstack([np.hstack(edge_from_face(f)) for f in fs])
-
-        # 3: Find the set of unique vertices in `edges`
-        v1s, v2s = np.hsplit(edges, 2)
-        verts = edges.reshape((-1, 3))
-        verts = np.vstack(sorted(verts, key=operator.itemgetter(0, 1, 2)))
-        eps = 1e-15 # the floating point calculation of the intersection locations is not _quite_ exact
-        verts = verts[list(np.sqrt(np.sum(np.diff(verts, axis=0) ** 2, axis=1)) > eps) + [True]]
-        # the True at the end there is because np.diff returns pairwise differences; one less element than the original array
+        class EdgeMap(object):
+            # A quick two level dictionary where the two keys are interchangeable (i.e. a symmetric graph)
+            def __init__(self):
+                self.d = {} # store indicies into self.values here, to make it easier to get inds or values
+                self.values = []
+            def _order(self, u, v):
```

Comment:
Here's another bit hotspot optimization. This method gets called a bazillion times, and it's faster to do a quick if statement than the more pythonic `sorted((u, v))` or a use of `min` & `max`. This optimization saved a second and a half.